Track times

rbenes opened the pull request:
I implement feature request https://github.com/pandas-dev/pandas/issues/32682 but if someone has more experience with pytables consider these changes.

The h5 files generated with track_times=False has the same md5 hashes, that is covered in the unit test as well as the fact that the hashes are not the same with track_times=True

- [x] closes #32682
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

rbenes committed:
Propagate track times

rbenes committed:
Make both tests

rbenes committed:
Unify

rbenes committed:
Note

jreback commented:
https://www.pytables.org/usersguide/libref/file_class.html?highlight=create_table#tables.File.create_table this option already defaults to `True`. is the purpose of this PR not to set it?

rbenes commented:
> https://www.pytables.org/usersguide/libref/file_class.html?highlight=create_table#tables.File.create_table this option already defaults to `True`. is the purpose of this PR not to set it?

Yes, you are right. The default value is `True` that means that creation time is stored inside h5 (that is reported as unwanted due to change of SHA hash). I exposed the track_times to `put` method of HDFStore. I let default value to `True` in all methods that it goes through to keep current behavior. Finally i pass it to `table._handle.create_table`.

The possibility to set of track_times can be seen in test on line 325. And on lines 340 and 343 you can see the different or the same check-sums after appropriate setup of track_times.

rbenes committed:
Black

rbenes committed:
Make it tables < 3.4.3 friendly

jreback commented on pandas/io/pytables.py:
```diff
@@ -4140,6 +4145,17 @@ def write(
+                from tables import __version__ as tables_version
+                from packaging.version import Version
+
+                if Version(tables_version) >= Version("3.4.3"):
```
use LooseVersion

rbenes replied:
changed. Thanks for hint. But, what is the reason for using LooseVersion? I found that `packaging.version` is more "modern" - https://stackoverflow.com/a/11887885/1386342

jreback commented on pandas/io/pytables.py:
```diff
@@ -4140,6 +4145,17 @@ def write(
 
+            if track_times is not None:
+                from tables import __version__ as tables_version
+                from packaging.version import Version
```
this import can be at the top (but use LooseVersion)

rbenes replied:
Moved. LooseVersion used.

rbenes replied:
I had to revert
```
from tables import __version__ as tables_version
```

 back here because of:

```
 Checks on imported code1s
##[error]Process completed with exit code 2.
Run source activate pandas-dev
Check import. No warnings, and blacklist some optional dependencies
err: pandas should not import: tables, numexpr
Check import. No warnings, and blacklist some optional dependencies DONE
##[error]Process completed with exit code 2.
```

jreback commented on pandas/io/pytables.py:
```diff
@@ -4140,6 +4145,17 @@ def write(
+
+                if Version(tables_version) >= Version("3.4.3"):
+                    options["track_times"] = track_times
+                else:
```
i would just ignore it

rbenes replied:
Yes. Ignored. It is now much easier.

jreback commented on pandas/io/pytables.py:
```diff
@@ -984,6 +984,7 @@ def put(
         data_columns: Optional[List[str]] = None,
         encoding=None,
         errors: str = "strict",
+        track_times: Optional[bool] = None,
```
default should be True

rbenes replied:
Chahged

jreback commented on pandas/tests/io/pytables/test_store.py:
```diff
@@ -296,6 +296,61 @@ def test_keys(self, setup_path):
+        # GH 32682
+        # enables to set track_times (see `pytables` `create_table` documentation)
+
+        import hashlib
```
all imports should be at the top; if you need to skip for a version you can do it with a `@td.skip_if_no`. add a test for >= 3.4.3 and one fore before that.

rbenes replied:
done :+1:

rbenes committed:
Use LooseVersion

rbenes committed:
Use default true. Be silent if unsupported pytables

rbenes committed:
Imports arranged

rbenes committed:
Mark test as skipif

rbenes committed:
Avoid importing tables

rbenes committed:
Revert import tables back to function

rbenes commented:
Hi @jreback . I did changes you requested. But, I found another possibility - not to pass track_times in parameters but put it here: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html . What do you thing?

jreback reviewed (changes requested):
we already require 3.4.2, so ok to simply bump the minimum to 3.4.3 is fine.

jreback commented:
> Hi @jreback . I did changes you requested. But, I found another possibility - not to pass track_times in parameters but put it here: https://pandas.pydata.org/pandas-docs/stable/user_guide/options.html . What do you think?

i think accepting it like in this pr is ok.

rbenes committed:
No need for handling of pytables version

rbenes committed:
Whats new added

rbenes commented:
> we already require 3.4.2, so ok to simply bump the minimum to 3.4.3 is fine.

I updated requirements in separate PR and updated this PR accordingly.

rbenes commented:
Hi, I see that "krysma" already remind this feature. Please revise changes and decide next steps. I synced it with master, so now it is without conflicts.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -220,19 +220,7 @@ Other enhancements
-  compression library. Compression was also added to the low-level Stata-file writers
-  :class:`~pandas.io.stata.StataWriter`, :class:`~pandas.io.stata.StataWriter117`,
-  and :class:`~pandas.io.stata.StataWriterUTF8` (:issue:`26599`).
+- :meth:`HDFStore.put` now accepts `track_times` parameter. Parameter is passed to ``create_table`` method of ``PyTables`` (:issue:`32682`).
```
don't change anything else, pls revert, then add this parameter after

rbenes replied:
o-o, first time I tried to solve conflict via github page and failed :-) ...
rebased and fixed

jreback commented on pandas/io/pytables.py:
```diff
@@ -984,6 +984,7 @@ def put(
         data_columns: Optional[List[str]] = None,
         encoding=None,
         errors: str = "strict",
+        track_times: bool = True,
```
can you update the doc-string, make sure to add a 1.1 versionadded tag

rbenes replied:
doc string added

jreback commented on pandas/io/pytables.py:
```diff
@@ -1027,6 +1028,7 @@ def put(
             data_columns=data_columns,
             encoding=encoding,
             errors=errors,
+            track_times=track_times,
```
same

rbenes replied:
same = you mean docstring? I do not see any docstring here.

jreback reviewed (changes requested):
looks fine, apart from comments

radekbenesinno committed:
Docstring added

jreback commented:
thanks @rbenes

Pierre-Bartet commented:
Hi, trying to use track_times=False I add to look at all the options used in the [test](https://github.com/pandas-dev/pandas/blob/b1b70c7390e589bbfa0d8896aa76e64bec0cf51e/pandas/tests/io/pytables/test_store.py#L301), so that I found that index=None is needed in order to indeed obtain reproducible files.
Is it expected? If yes it should be said somewhere in the documentation.
